// drag-timetable/dist/task.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _util = require('./util.js');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableTask = function () {
  function TimetableTask(task, taskAreaSize) {
    (0, _classCallCheck3.default)(this, TimetableTask);

    this.id = task.id;
    this.start = task.start;
    this.end = task.end;
    this.text = task.text;
    this._initTaskElement(taskAreaSize);
  }

  (0, _createClass3.default)(TimetableTask, [{
    key: '_initTaskElement',
    value: function _initTaskElement(taskAreaSize) {
      this.element = document.createElement('div');
      this.actualTask = document.createElement('div');
      this.actualTask.setAttribute('class', 'dragtimetable-task');
      this.element.style.width = '100%';
      this.element.style.height = '100%';
      this.element.style.boxSizing = 'border-box';
      this.element.style.paddingLeft = '10px';
      this.element.style.paddingRight = '10px';
      this.element.style.position = 'relative';

      this.updateTaskUI();

      this.element.appendChild(this.actualTask);
    }
  }, {
    key: 'updateTaskUI',
    value: function updateTaskUI() {
      if (this.text) {
        this.actualTask.innerHTML = this.text;
      }

      this.actualTask.appendChild(this.getTimeElement());
    }
  }, {
    key: 'getTimeElement',
    value: function getTimeElement() {
      var timeElement = document.createElement('div');
      timeElement.innerHTML = '&#x1f550; ' + _util2.default.getHourText(this.start) + ' - ' + _util2.default.getHourText(this.end);
      timeElement.style.position = 'absolute';
      timeElement.style.bottom = '0px';
      timeElement.style.paddingBottom = '3px';
      return timeElement;
    }
  }]);
  return TimetableTask;
}();

exports.default = TimetableTask;
