// drag-timetable/demo/index.build.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
window.onload = function() {
  var DragTimetable = require('../dist/app.js');

  var instance = DragTimetable.create(document.getElementById("timetable_container"), { hourStart: 8, hourEnd: 14 });

  instance.addTask({id: 1, start: 9, end: 11, text: 'Task 1 - Get this stuff done!'}, true);
  instance.addTask({id: 2, start: 12, end: 13, text: 'Task 2 - Get that stuff done!'}, true);

  instance.setMoveCallback(function(task) {
    console.log('task moved: ' + task.id);
  });

  instance.setClickCallback(function(task) {
    console.log('task clicked: ' + task.id);
  });
};

},{"../dist/app.js":2}],2:[function(require,module,exports){
'use strict';

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _create = require('./create.js');

var _create2 = _interopRequireDefault(_create);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DragTimetable = function () {
  function DragTimetable() {
    (0, _classCallCheck3.default)(this, DragTimetable);
  }

  (0, _createClass3.default)(DragTimetable, [{
    key: 'create',
    value: function create(container, options) {
      if (!options || !container) {
        throw 'No options or container';
      }

      var optionsObj = {};

      //horizontal or vertical
      optionsObj.orientation = 'vertical';
      if (options.orientation) {
        optionsObj.orientation = options.orientation;
      }

      // 0 to 23
      optionsObj.hourStart = 8;
      if (options.hourStart !== undefined) {
        optionsObj.hourStart = options.hourStart;
      }

      // 0 to 23
      optionsObj.hourEnd = 16;
      if (options.hourEnd !== undefined) {
        optionsObj.hourEnd = options.hourEnd;
      }

      // 0 to 6
      optionsObj.dayStart = 1;
      if (options.dayStart !== undefined) {
        optionsObj.dayStart = options.dayStart;
      }

      // 0 to 6
      optionsObj.dayEnd = 5;
      if (options.dayEnd !== undefined) {
        optionsObj.dayEnd = options.dayEnd;
      }

      //day or week (maybe month in the future?)
      optionsObj.timeMode = 'day';
      if (options.timeMode) {
        optionsObj.timeMode = options.timeMode;
      }

      //any css size unit
      optionsObj.timeHeaderSize = '100px';
      if (options.timeHeaderSize) {
        optionsObj.timeHeaderSize = options.timeHeaderSize;
      }

      optionsObj.taskAreaSize = '150px';
      if (options.taskAreaSize) {
        optionsObj.taskAreaSize = options.taskAreaSize;
      }

      optionsObj.quarterHourAreaSize = '17px';
      if (options.quarterHourAreaSize) {
        optionsObj.quarterHourAreaSize = options.quarterHourAreaSize;
      }

      optionsObj.is24Clock = false;
      if (options.is24Clock) {
        optionsObj.is24Clock = options.is24Clock;
      }

      optionsObj.clickThreshold = 200;
      if (options.clickThreshold) {
        optionsObj.clickThreshold = options.clickThreshold;
      }

      return _create2.default.create(container, optionsObj);
    }
  }]);
  return DragTimetable;
}();

module.exports = new DragTimetable();
},{"./create.js":3,"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],3:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _instance = require('./instance.js');

var _instance2 = _interopRequireDefault(_instance);

var _spacer = require('./spacer.js');

var _spacer2 = _interopRequireDefault(_spacer);

var _drag = require('./drag.js');

var _drag2 = _interopRequireDefault(_drag);

var _util = require('./util.js');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableCreator = function () {
  function TimetableCreator() {
    (0, _classCallCheck3.default)(this, TimetableCreator);
  }

  (0, _createClass3.default)(TimetableCreator, [{
    key: 'create',
    value: function create(container, options) {
      var hoursArray = this._getHoursArray(options.hourStart, options.hourEnd);

      var timetableContainer = document.createElement('div');
      timetableContainer.style.position = 'relative';

      options.hoursArray = hoursArray;

      var tableResult = this._createTable(options);

      timetableContainer.appendChild(tableResult.table);
      timetableContainer.appendChild(tableResult.taskArea);
      container.appendChild(timetableContainer);

      var contextObj = (0, _assign2.default)(tableResult, options);
      contextObj.tasks = {};

      var instance = new _instance2.default(contextObj);
      this._initSpacer(instance);
      this._initDrag(instance);
      return instance;
    }
  }, {
    key: '_createTable',
    value: function _createTable(options) {
      var table = document.createElement('table');

      table.setAttribute('class', 'dragtimetable-table');

      options.table = table;
      var taskArea = this._createVerticalTable(options);
      return { table: table, taskArea: taskArea };
    }
  }, {
    key: '_createVerticalTable',
    value: function _createVerticalTable(_ref) {
      var orientation = _ref.orientation,
          table = _ref.table,
          hoursArray = _ref.hoursArray,
          quarterHourAreaSize = _ref.quarterHourAreaSize,
          timeHeaderSize = _ref.timeHeaderSize,
          taskAreaSize = _ref.taskAreaSize;

      var taskarea = document.createElement('table');
      taskarea.style.position = 'absolute';
      taskarea.style.top = '0px';
      taskarea.style.left = timeHeaderSize;
      taskarea.style.boxSizing = 'border-box';
      taskarea.style.marginLeft = '4px';
      taskarea.style.width = taskAreaSize;
      taskarea.style.borderSpacing = '1px';
      taskarea.style.borderCollapse = 'collapse';

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = (0, _getIterator3.default)(hoursArray), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var hour = _step.value;

          var hourBlock = document.createElement('tr');
          var hourHeader = document.createElement('th');
          var hourHeaderText = document.createElement('div');
          var taskCell = document.createElement('td');

          hourHeaderText.innerHTML = hour;
          hourHeader.style.width = timeHeaderSize;
          hourHeader.style.padding = '0px';
          hourHeader.setAttribute('rowspan', '4');
          hourHeaderText.setAttribute("class", "dragtimetable-headertext");
          hourHeader.appendChild(hourHeaderText);
          hourBlock.appendChild(hourHeader);
          taskCell.style.width = taskAreaSize;
          taskCell.style.padding = '0px';
          hourBlock.style.height = quarterHourAreaSize;
          hourBlock.appendChild(taskCell);
          table.appendChild(hourBlock);

          for (var i = 0; i < 3; i++) {
            var taskSplitRow = document.createElement('tr');
            var taskSplitCell = document.createElement('td');
            taskSplitCell.style.padding = '0px';
            taskSplitRow.style.height = quarterHourAreaSize;
            taskSplitRow.appendChild(taskSplitCell);
            table.appendChild(taskSplitRow);
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return taskarea;
    }
  }, {
    key: '_getHoursArray',
    value: function _getHoursArray(hourStart, hourEnd) {
      var hours = [];
      if (hourStart < hourEnd) {
        for (var i = hourStart; i < hourEnd; i++) {
          hours.push(_util2.default.getHourText(i));
        }
      } else {
        for (var _i = hourStart; _i < 24; _i++) {
          hours.push(_util2.default.getHourText(_i));
        }
        for (var _i2 = 0; _i2 < hourEnd; _i2++) {
          hours.push(_util2.default.getHourText(_i2));
        }
      }
      return hours;
    }
  }, {
    key: '_initSpacer',
    value: function _initSpacer(instance) {
      var callbacks = {
        onTaskTimeUpdate: instance.onTaskTimeUpdate.bind(instance),
        getTask: instance.getTask.bind(instance)
      };
      instance.contextObj.spacer = new _spacer2.default(instance.contextObj.hourStart, instance.contextObj.hourEnd, instance.contextObj.taskArea, instance.contextObj.table, instance.contextObj.quarterHourAreaSize, instance.contextObj.orientation, callbacks);
    }
  }, {
    key: '_initDrag',
    value: function _initDrag(instance) {
      var callbacks = {
        onTaskClick: instance.onTaskClick.bind(instance)
      };
      instance.contextObj.dragManager = new _drag2.default(instance.quarterHourAreaSize, instance.contextObj.spacer, callbacks, instance.contextObj.clickThreshold);
    }
  }]);
  return TimetableCreator;
}();

exports.default = new TimetableCreator();
},{"./drag.js":4,"./instance.js":5,"./spacer.js":6,"./util.js":8,"babel-runtime/core-js/get-iterator":9,"babel-runtime/core-js/object/assign":10,"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],4:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableDraggable = function () {
  function TimetableDraggable(taskId, element, dragManager, callbacks) {
    (0, _classCallCheck3.default)(this, TimetableDraggable);

    this.taskId = taskId;
    this.element = element;
    this.dragManager = dragManager;
    this.isDragging = false;
    this.element.onmousedown = this.onTouchStart.bind(this);
    window.addEventListener('mousemove', this.onTouchMove.bind(this), false);
    window.addEventListener('mouseup', this.onTouchEnd.bind(this), false);
  }

  (0, _createClass3.default)(TimetableDraggable, [{
    key: 'onTouchStart',
    value: function onTouchStart(ev) {
      this.isDragging = true;
      var boundingRect = ev.currentTarget.getBoundingClientRect();
      this.startOffsetX = ev.clientX - boundingRect.left;
      this.startOffsetY = ev.clientY - boundingRect.top;
      this.startTouchTime = new Date().getTime();
      this.currentGhostElement = this.element.cloneNode(true);
      this.currentGhostElement.style.position = 'absolute';
      this.currentGhostElement.style.left = ev.pageX - this.startOffsetX + "px";
      this.currentGhostElement.style.top = ev.pageY - this.startOffsetY + "px";
      this.currentGhostElement.style.width = this.element.clientWidth + "px";
      this.currentGhostElement.style.height = this.element.clientHeight + "px";
      document.body.appendChild(this.currentGhostElement);
      this.dragManager.spacer.moveStart(this.taskId);
      return false;
    }
  }, {
    key: 'onTouchEnd',
    value: function onTouchEnd(ev) {
      if (this.isDragging) {
        this.isDragging = false;
        if (new Date().getTime() - this.startTouchTime < this.dragManager.clickThreshold) {
          this.dragManager.callbacks.onTaskClick(this.taskId);
        }
        if (this.currentGhostElement) {
          this.currentGhostElement.outerHTML = '';
          this.currentGhostElement = null;
        }
        this.dragManager.spacer.moveEnd(this.taskId);
      }
    }
  }, {
    key: 'onTouchMove',
    value: function onTouchMove(ev) {
      if (this.isDragging) {
        if (this.currentGhostElement) {
          this.currentGhostElement.style.left = ev.pageX - this.startOffsetX + "px";
          this.currentGhostElement.style.top = ev.pageY - this.startOffsetY + "px";
        }
        var scrollOffset = window.pageYOffset || document.documentElement.scrollTop;
        this.dragManager.spacer.moveUpdate(ev.pageY - this.startOffsetY - scrollOffset, this.taskId);
      }
    }
  }]);
  return TimetableDraggable;
}();

var TimetableDragManager = function () {
  function TimetableDragManager(unitHeight, spacer, callbacks, clickThreshold) {
    (0, _classCallCheck3.default)(this, TimetableDragManager);

    this.callbacks = callbacks;
    this.unitHeight = unitHeight;
    this.spacer = spacer;
    this.clickThreshold = clickThreshold;
    this.taskElements = {};
  }

  (0, _createClass3.default)(TimetableDragManager, [{
    key: 'attachToTaskElement',
    value: function attachToTaskElement(taskId, element) {
      var draggable = new TimetableDraggable(taskId, element, this);
      this.taskElements[taskId] = draggable;
    }
  }, {
    key: 'detachToTaskElement',
    value: function detachToTaskElement(taskId, element) {}
  }]);
  return TimetableDragManager;
}();

exports.default = TimetableDragManager;
},{"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],5:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _task = require('./task.js');

var _task2 = _interopRequireDefault(_task);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableInstance = function () {
  function TimetableInstance(contextObj) {
    (0, _classCallCheck3.default)(this, TimetableInstance);

    this.contextObj = contextObj;
  }

  (0, _createClass3.default)(TimetableInstance, [{
    key: 'setMoveCallback',
    value: function setMoveCallback(moveCallback) {
      this.moveCallback = moveCallback;
    }
  }, {
    key: 'setClickCallback',
    value: function setClickCallback(clickCallback) {
      this.clickCallback = clickCallback;
    }
  }, {
    key: 'addTask',
    value: function addTask(task, isAddingToTimetable) {
      var newTask = new _task2.default(task, this.contextObj.taskAreaSize);
      this.contextObj.tasks[task.id] = newTask;
      this.contextObj.dragManager.attachToTaskElement(newTask.id, newTask.element);
      if (isAddingToTimetable) {
        this.contextObj.spacer.addToTable(newTask);
      }
      return newTask.element;
    }
  }, {
    key: 'removeTask',
    value: function removeTask(taskId) {
      this.contextObj.spacer.removeFromTable(this.contextObj.tasks[taskId]);
      this.contextObj.tasks[taskId] = undefined;
    }
  }, {
    key: 'onTaskTimeUpdate',
    value: function onTaskTimeUpdate(taskId, startTime, endTime) {
      this.contextObj.tasks[taskId].start = startTime;
      this.contextObj.tasks[taskId].end = endTime;
      if (this.moveCallback) {
        this.moveCallback(this.contextObj.tasks[taskId]);
      }
      this.contextObj.tasks[taskId].updateTaskUI();
    }
  }, {
    key: 'onTaskClick',
    value: function onTaskClick(taskId) {
      if (this.clickCallback) {
        this.clickCallback(this.contextObj.tasks[taskId]);
      }
    }
  }, {
    key: 'getTask',
    value: function getTask(taskId) {
      return this.contextObj.tasks[taskId];
    }
  }, {
    key: 'getTasks',
    value: function getTasks() {
      return this.contextObj.tasks;
    }
  }]);
  return TimetableInstance;
}();

exports.default = TimetableInstance;
},{"./task.js":7,"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],6:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableSpacer = function () {
  function TimetableSpacer(hourStart, hourEnd, taskArea, table, unitHeight, orientation, callbacks) {
    (0, _classCallCheck3.default)(this, TimetableSpacer);

    this.hourStart = hourStart;
    this.hourEnd = hourEnd;
    this.taskArea = taskArea;
    this.table = table;
    this.unitHeight = unitHeight;
    this.orientation = orientation;
    this.callbacks = callbacks;
    this.spacerIdToElementArray = [];
    this.timeToSpacerMap = {};
    this.spacerToTimeMap = {};
    this._initSpacers();
  }

  (0, _createClass3.default)(TimetableSpacer, [{
    key: 'moveStart',
    value: function moveStart(selectedTaskId) {
      var selectedTask = this.callbacks.getTask(selectedTaskId);
      var boundary = this.timeToSpacerMap[selectedTask.start].boundary;
      this.currentAvailSlot = { spacerId: boundary[0], lastSpacerId: boundary[1] };
      this.removeFromTable(selectedTask);
    }
  }, {
    key: 'moveUpdate',
    value: function moveUpdate(mouseOffset, selectedTaskId) {
      var selectedTask = this.callbacks.getTask(selectedTaskId);
      var availSlotResult = this.getClosestAvailableSlot(mouseOffset, selectedTask);
      if (availSlotResult) {
        this.currentAvailSlot = availSlotResult;
      }
    }
  }, {
    key: 'moveEnd',
    value: function moveEnd(selectedTaskId) {
      var selectedTask = this.callbacks.getTask(selectedTaskId);
      if (!this.currentAvailSlot) {
        return;
      }
      this.addToTable(selectedTask, this.currentAvailSlot.spacerId, this.currentAvailSlot.lastSpacerId);
      this.callbacks.onTaskTimeUpdate(selectedTask.id, this.spacerToTimeMap[this.currentAvailSlot.spacerId], this.spacerToTimeMap[this.currentAvailSlot.lastSpacerId] + 0.25);
    }
  }, {
    key: 'getClosestAvailableSlot',
    value: function getClosestAvailableSlot(mouseOffset, selectedTask) {
      var spacerLength = Math.floor((selectedTask.end - selectedTask.start) * 4.0);
      for (var i = this.spacerIdToElementArray.length - 1; i >= 0; i--) {
        var spacerElement = this.spacerIdToElementArray[i];
        if (spacerElement.style.display === 'none') {
          continue;
        }
        var clientRect = spacerElement.getBoundingClientRect();
        var spacerPosition = this.orientation === 'horizontal' ? clientRect.left : clientRect.top;
        if (i == 0 || mouseOffset > spacerPosition) {
          var conflictDetected = false;
          var lastSpacerId = 0;
          for (var f = i; f < i + spacerLength; f += 1) {
            var currTime = this.spacerToTimeMap[f];
            if (this.timeToSpacerMap[currTime] === undefined || this.timeToSpacerMap[currTime] === null || this.timeToSpacerMap[currTime] && this.timeToSpacerMap[currTime].taskId && this.timeToSpacerMap[currTime].taskId !== selectedTask.id) {
              conflictDetected = true;
            }
            lastSpacerId = f;
          }
          if (!conflictDetected) {
            return { spacerId: i, lastSpacerId: lastSpacerId };
          }
        }
      }
      return false;
    }
  }, {
    key: 'addToTable',
    value: function addToTable(task, availSlotStart, availSlotEnd) {
      var taskStart = task.start;
      var taskEnd = task.end;
      if (availSlotStart !== undefined && availSlotStart !== null && availSlotEnd !== undefined && availSlotEnd !== null) {
        taskStart = this.spacerToTimeMap[availSlotStart];
        taskEnd = this.spacerToTimeMap[availSlotEnd + 1];
      }

      for (var i = 0.0 + taskStart; i < taskEnd; i += 0.25) {
        if (this.timeToSpacerMap[i] && this.timeToSpacerMap[i].taskId) {
          throw 'No available slots!';
        }
      }
      var startSpacerId = this.timeToSpacerMap[taskStart];
      var endSpacerId = this.timeToSpacerMap[taskEnd - 0.25];
      for (var _i = 0.0 + taskStart; _i < taskEnd; _i += 0.25) {
        var spacerId = this.timeToSpacerMap[_i];
        this.timeToSpacerMap[_i] = { taskId: task.id, boundary: [this.timeToSpacerMap[taskStart], this.timeToSpacerMap[taskEnd - 0.25]], availableValue: spacerId };
        if (_i == taskStart) {
          this.spacerIdToElementArray[spacerId].style.display = '';
          this.spacerIdToElementArray[spacerId].appendChild(task.element);
          this.spacerIdToElementArray[spacerId].setAttribute('rowspan', '' + (endSpacerId - startSpacerId + 1));
        } else {
          this.spacerIdToElementArray[spacerId].style.display = 'none';
        }
      }
    }
  }, {
    key: 'removeFromTable',
    value: function removeFromTable(task) {
      var startSpacerId = this.timeToSpacerMap[task.start];
      var endSpacerId = this.timeToSpacerMap[task.end - 0.25];
      for (var i = 0.0 + task.start; i < task.end; i += 0.25) {
        var spacerObj = this.timeToSpacerMap[i];
        if (!spacerObj.taskId) {
          continue;
        }
        this.timeToSpacerMap[i] = spacerObj.availableValue;
        var spacerId = spacerObj.availableValue;
        if (i == task.start) {
          this.spacerIdToElementArray[spacerId].removeChild(this.spacerIdToElementArray[spacerId].firstChild);
          this.spacerIdToElementArray[spacerId].setAttribute('rowspan', '1');
        } else {
          this.spacerIdToElementArray[spacerId].style.display = '';
        }
      }
    }
  }, {
    key: '_initSpacers',
    value: function _initSpacers() {
      var _this = this;

      var isFirefox = navigator.userAgent.toLowerCase().indexOf('firefox') > -1;

      for (var i = 0.0 + this.hourStart; i < this.hourEnd; i += 0.25) {
        var rowElement = document.createElement('tr');
        var spaceElement = document.createElement('td');
        rowElement.style.height = this.unitHeight;
        spaceElement.setAttribute('class', 'dragtimetable-spacer');
        if (isFirefox) {
          spaceElement.style.height = '100%';
        }
        this.spacerIdToElementArray.push(spaceElement);
        this.timeToSpacerMap[i] = this.spacerIdToElementArray.length - 1;
        this.spacerToTimeMap[this.spacerIdToElementArray.length - 1] = i;
        rowElement.appendChild(spaceElement);
        this.taskArea.appendChild(rowElement);
      }
      this.spacerToTimeMap[this.spacerIdToElementArray.length] = this.hourEnd;
      window.addEventListener('resize', function () {
        _this.taskArea.style.height = _this.table.clientHeight + "px";
      });
      this.taskArea.style.height = this.table.clientHeight + "px";
    }
  }]);
  return TimetableSpacer;
}();

exports.default = TimetableSpacer;
},{"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],7:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _util = require('./util.js');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableTask = function () {
  function TimetableTask(task, taskAreaSize) {
    (0, _classCallCheck3.default)(this, TimetableTask);

    this.id = task.id;
    this.start = task.start;
    this.end = task.end;
    this.text = task.text;
    this._initTaskElement(taskAreaSize);
  }

  (0, _createClass3.default)(TimetableTask, [{
    key: '_initTaskElement',
    value: function _initTaskElement(taskAreaSize) {
      this.element = document.createElement('div');
      this.actualTask = document.createElement('div');
      this.actualTask.setAttribute('class', 'dragtimetable-task');
      this.element.style.width = '100%';
      this.element.style.height = '100%';
      this.element.style.boxSizing = 'border-box';
      this.element.style.paddingLeft = '10px';
      this.element.style.paddingRight = '10px';
      this.element.style.position = 'relative';

      this.updateTaskUI();

      this.element.appendChild(this.actualTask);
    }
  }, {
    key: 'updateTaskUI',
    value: function updateTaskUI() {
      if (this.text) {
        this.actualTask.innerHTML = this.text;
      }

      this.actualTask.appendChild(this.getTimeElement());
    }
  }, {
    key: 'getTimeElement',
    value: function getTimeElement() {
      var timeElement = document.createElement('div');
      timeElement.innerHTML = '&#x1f550; ' + _util2.default.getHourText(this.start) + ' - ' + _util2.default.getHourText(this.end);
      timeElement.style.position = 'absolute';
      timeElement.style.bottom = '0px';
      timeElement.style.paddingBottom = '3px';
      return timeElement;
    }
  }]);
  return TimetableTask;
}();

exports.default = TimetableTask;
},{"./util.js":8,"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],8:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TimetableUtil = function () {
  function TimetableUtil() {
    (0, _classCallCheck3.default)(this, TimetableUtil);
  }

  (0, _createClass3.default)(TimetableUtil, [{
    key: 'getHourText',
    value: function getHourText(i) {
      var minuteMap = {
        0.25: ':15',
        0.50: ':30',
        0.75: ':45'
      };
      var minute = minuteMap[i % 1.0] ? minuteMap[i % 1.0] : '';
      var fi = Math.floor(i);
      return i == 0 ? "12 AM" : fi >= 12 ? (fi == 12 ? 12 : fi - 12) + minute + " PM" : fi + minute + " AM";
    }
  }]);
  return TimetableUtil;
}();

exports.default = new TimetableUtil();
},{"babel-runtime/helpers/classCallCheck":12,"babel-runtime/helpers/createClass":13}],9:[function(require,module,exports){
module.exports = { "default": require("core-js/library/fn/get-iterator"), __esModule: true };
},{"core-js/library/fn/get-iterator":14}],10:[function(require,module,exports){
module.exports = { "default": require("core-js/library/fn/object/assign"), __esModule: true };
},{"core-js/library/fn/object/assign":15}],11:[function(require,module,exports){
module.exports = { "default": require("core-js/library/fn/object/define-property"), __esModule: true };
},{"core-js/library/fn/object/define-property":16}],12:[function(require,module,exports){
"use strict";

exports.__esModule = true;

exports.default = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};
},{}],13:[function(require,module,exports){
"use strict";

exports.__esModule = true;

var _defineProperty = require("../core-js/object/define-property");

var _defineProperty2 = _interopRequireDefault(_defineProperty);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      (0, _defineProperty2.default)(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();
},{"../core-js/object/define-property":11}],14:[function(require,module,exports){
require('../modules/web.dom.iterable');
require('../modules/es6.string.iterator');
module.exports = require('../modules/core.get-iterator');

},{"../modules/core.get-iterator":67,"../modules/es6.string.iterator":71,"../modules/web.dom.iterable":72}],15:[function(require,module,exports){
require('../../modules/es6.object.assign');
module.exports = require('../../modules/_core').Object.assign;

},{"../../modules/_core":23,"../../modules/es6.object.assign":69}],16:[function(require,module,exports){
require('../../modules/es6.object.define-property');
var $Object = require('../../modules/_core').Object;
module.exports = function defineProperty(it, key, desc) {
  return $Object.defineProperty(it, key, desc);
};

},{"../../modules/_core":23,"../../modules/es6.object.define-property":70}],17:[function(require,module,exports){
module.exports = function (it) {
  if (typeof it != 'function') throw TypeError(it + ' is not a function!');
  return it;
};

},{}],18:[function(require,module,exports){
module.exports = function () { /* empty */ };

},{}],19:[function(require,module,exports){
var isObject = require('./_is-object');
module.exports = function (it) {
  if (!isObject(it)) throw TypeError(it + ' is not an object!');
  return it;
};

},{"./_is-object":37}],20:[function(require,module,exports){
// false -> Array#indexOf
// true  -> Array#includes
var toIObject = require('./_to-iobject');
var toLength = require('./_to-length');
var toAbsoluteIndex = require('./_to-absolute-index');
module.exports = function (IS_INCLUDES) {
  return function ($this, el, fromIndex) {
    var O = toIObject($this);
    var length = toLength(O.length);
    var index = toAbsoluteIndex(fromIndex, length);
    var value;
    // Array#includes uses SameValueZero equality algorithm
    // eslint-disable-next-line no-self-compare
    if (IS_INCLUDES && el != el) while (length > index) {
      value = O[index++];
      // eslint-disable-next-line no-self-compare
      if (value != value) return true;
    // Array#indexOf ignores holes, Array#includes - not
    } else for (;length > index; index++) if (IS_INCLUDES || index in O) {
      if (O[index] === el) return IS_INCLUDES || index || 0;
    } return !IS_INCLUDES && -1;
  };
};

},{"./_to-absolute-index":58,"./_to-iobject":60,"./_to-length":61}],21:[function(require,module,exports){
// getting tag from 19.1.3.6 Object.prototype.toString()
var cof = require('./_cof');
var TAG = require('./_wks')('toStringTag');
// ES3 wrong here
var ARG = cof(function () { return arguments; }()) == 'Arguments';

// fallback for IE11 Script Access Denied error
var tryGet = function (it, key) {
  try {
    return it[key];
  } catch (e) { /* empty */ }
};

module.exports = function (it) {
  var O, T, B;
  return it === undefined ? 'Undefined' : it === null ? 'Null'
    // @@toStringTag case
    : typeof (T = tryGet(O = Object(it), TAG)) == 'string' ? T
    // builtinTag case
    : ARG ? cof(O)
    // ES3 arguments fallback
    : (B = cof(O)) == 'Object' && typeof O.callee == 'function' ? 'Arguments' : B;
};

},{"./_cof":22,"./_wks":65}],22:[function(require,module,exports){
var toString = {}.toString;

module.exports = function (it) {
  return toString.call(it).slice(8, -1);
};

},{}],23:[function(require,module,exports){
var core = module.exports = { version: '2.5.3' };
if (typeof __e == 'number') __e = core; // eslint-disable-line no-undef

},{}],24:[function(require,module,exports){
// optional / simple context binding
var aFunction = require('./_a-function');
module.exports = function (fn, that, length) {
  aFunction(fn);
  if (that === undefined) return fn;
  switch (length) {
    case 1: return function (a) {
      return fn.call(that, a);
    };
    case 2: return function (a, b) {
      return fn.call(that, a, b);
    };
    case 3: return function (a, b, c) {
      return fn.call(that, a, b, c);
    };
  }
  return function (/* ...args */) {
    return fn.apply(that, arguments);
  };
};

},{"./_a-function":17}],25:[function(require,module,exports){
// 7.2.1 RequireObjectCoercible(argument)
module.exports = function (it) {
  if (it == undefined) throw TypeError("Can't call method on  " + it);
  return it;
};

},{}],26:[function(require,module,exports){
// Thank's IE8 for his funny defineProperty
module.exports = !require('./_fails')(function () {
  return Object.defineProperty({}, 'a', { get: function () { return 7; } }).a != 7;
});

},{"./_fails":30}],27:[function(require,module,exports){
var isObject = require('./_is-object');
var document = require('./_global').document;
// typeof document.createElement is 'object' in old IE
var is = isObject(document) && isObject(document.createElement);
module.exports = function (it) {
  return is ? document.createElement(it) : {};
};

},{"./_global":31,"./_is-object":37}],28:[function(require,module,exports){
// IE 8- don't enum bug keys
module.exports = (
  'constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf'
).split(',');

},{}],29:[function(require,module,exports){
var global = require('./_global');
var core = require('./_core');
var ctx = require('./_ctx');
var hide = require('./_hide');
var PROTOTYPE = 'prototype';

var $export = function (type, name, source) {
  var IS_FORCED = type & $export.F;
  var IS_GLOBAL = type & $export.G;
  var IS_STATIC = type & $export.S;
  var IS_PROTO = type & $export.P;
  var IS_BIND = type & $export.B;
  var IS_WRAP = type & $export.W;
  var exports = IS_GLOBAL ? core : core[name] || (core[name] = {});
  var expProto = exports[PROTOTYPE];
  var target = IS_GLOBAL ? global : IS_STATIC ? global[name] : (global[name] || {})[PROTOTYPE];
  var key, own, out;
  if (IS_GLOBAL) source = name;
  for (key in source) {
    // contains in native
    own = !IS_FORCED && target && target[key] !== undefined;
    if (own && key in exports) continue;
    // export native or passed
    out = own ? target[key] : source[key];
    // prevent global pollution for namespaces
    exports[key] = IS_GLOBAL && typeof target[key] != 'function' ? source[key]
    // bind timers to global for call from export context
    : IS_BIND && own ? ctx(out, global)
    // wrap global constructors for prevent change them in library
    : IS_WRAP && target[key] == out ? (function (C) {
      var F = function (a, b, c) {
        if (this instanceof C) {
          switch (arguments.length) {
            case 0: return new C();
            case 1: return new C(a);
            case 2: return new C(a, b);
          } return new C(a, b, c);
        } return C.apply(this, arguments);
      };
      F[PROTOTYPE] = C[PROTOTYPE];
      return F;
    // make static versions for prototype methods
    })(out) : IS_PROTO && typeof out == 'function' ? ctx(Function.call, out) : out;
    // export proto methods to core.%CONSTRUCTOR%.methods.%NAME%
    if (IS_PROTO) {
      (exports.virtual || (exports.virtual = {}))[key] = out;
      // export proto methods to core.%CONSTRUCTOR%.prototype.%NAME%
      if (type & $export.R && expProto && !expProto[key]) hide(expProto, key, out);
    }
  }
};
// type bitmap
$export.F = 1;   // forced
$export.G = 2;   // global
$export.S = 4;   // static
$export.P = 8;   // proto
$export.B = 16;  // bind
$export.W = 32;  // wrap
$export.U = 64;  // safe
$export.R = 128; // real proto method for `library`
module.exports = $export;

},{"./_core":23,"./_ctx":24,"./_global":31,"./_hide":33}],30:[function(require,module,exports){
module.exports = function (exec) {
  try {
    return !!exec();
  } catch (e) {
    return true;
  }
};

},{}],31:[function(require,module,exports){
// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
var global = module.exports = typeof window != 'undefined' && window.Math == Math
  ? window : typeof self != 'undefined' && self.Math == Math ? self
  // eslint-disable-next-line no-new-func
  : Function('return this')();
if (typeof __g == 'number') __g = global; // eslint-disable-line no-undef

},{}],32:[function(require,module,exports){
var hasOwnProperty = {}.hasOwnProperty;
module.exports = function (it, key) {
  return hasOwnProperty.call(it, key);
};

},{}],33:[function(require,module,exports){
var dP = require('./_object-dp');
var createDesc = require('./_property-desc');
module.exports = require('./_descriptors') ? function (object, key, value) {
  return dP.f(object, key, createDesc(1, value));
} : function (object, key, value) {
  object[key] = value;
  return object;
};

},{"./_descriptors":26,"./_object-dp":45,"./_property-desc":52}],34:[function(require,module,exports){
var document = require('./_global').document;
module.exports = document && document.documentElement;

},{"./_global":31}],35:[function(require,module,exports){
module.exports = !require('./_descriptors') && !require('./_fails')(function () {
  return Object.defineProperty(require('./_dom-create')('div'), 'a', { get: function () { return 7; } }).a != 7;
});

},{"./_descriptors":26,"./_dom-create":27,"./_fails":30}],36:[function(require,module,exports){
// fallback for non-array-like ES3 and non-enumerable old V8 strings
var cof = require('./_cof');
// eslint-disable-next-line no-prototype-builtins
module.exports = Object('z').propertyIsEnumerable(0) ? Object : function (it) {
  return cof(it) == 'String' ? it.split('') : Object(it);
};

},{"./_cof":22}],37:[function(require,module,exports){
module.exports = function (it) {
  return typeof it === 'object' ? it !== null : typeof it === 'function';
};

},{}],38:[function(require,module,exports){
'use strict';
var create = require('./_object-create');
var descriptor = require('./_property-desc');
var setToStringTag = require('./_set-to-string-tag');
var IteratorPrototype = {};

// 25.1.2.1.1 %IteratorPrototype%[@@iterator]()
require('./_hide')(IteratorPrototype, require('./_wks')('iterator'), function () { return this; });

module.exports = function (Constructor, NAME, next) {
  Constructor.prototype = create(IteratorPrototype, { next: descriptor(1, next) });
  setToStringTag(Constructor, NAME + ' Iterator');
};

},{"./_hide":33,"./_object-create":44,"./_property-desc":52,"./_set-to-string-tag":54,"./_wks":65}],39:[function(require,module,exports){
'use strict';
var LIBRARY = require('./_library');
var $export = require('./_export');
var redefine = require('./_redefine');
var hide = require('./_hide');
var has = require('./_has');
var Iterators = require('./_iterators');
var $iterCreate = require('./_iter-create');
var setToStringTag = require('./_set-to-string-tag');
var getPrototypeOf = require('./_object-gpo');
var ITERATOR = require('./_wks')('iterator');
var BUGGY = !([].keys && 'next' in [].keys()); // Safari has buggy iterators w/o `next`
var FF_ITERATOR = '@@iterator';
var KEYS = 'keys';
var VALUES = 'values';

var returnThis = function () { return this; };

module.exports = function (Base, NAME, Constructor, next, DEFAULT, IS_SET, FORCED) {
  $iterCreate(Constructor, NAME, next);
  var getMethod = function (kind) {
    if (!BUGGY && kind in proto) return proto[kind];
    switch (kind) {
      case KEYS: return function keys() { return new Constructor(this, kind); };
      case VALUES: return function values() { return new Constructor(this, kind); };
    } return function entries() { return new Constructor(this, kind); };
  };
  var TAG = NAME + ' Iterator';
  var DEF_VALUES = DEFAULT == VALUES;
  var VALUES_BUG = false;
  var proto = Base.prototype;
  var $native = proto[ITERATOR] || proto[FF_ITERATOR] || DEFAULT && proto[DEFAULT];
  var $default = (!BUGGY && $native) || getMethod(DEFAULT);
  var $entries = DEFAULT ? !DEF_VALUES ? $default : getMethod('entries') : undefined;
  var $anyNative = NAME == 'Array' ? proto.entries || $native : $native;
  var methods, key, IteratorPrototype;
  // Fix native
  if ($anyNative) {
    IteratorPrototype = getPrototypeOf($anyNative.call(new Base()));
    if (IteratorPrototype !== Object.prototype && IteratorPrototype.next) {
      // Set @@toStringTag to native iterators
      setToStringTag(IteratorPrototype, TAG, true);
      // fix for some old engines
      if (!LIBRARY && !has(IteratorPrototype, ITERATOR)) hide(IteratorPrototype, ITERATOR, returnThis);
    }
  }
  // fix Array#{values, @@iterator}.name in V8 / FF
  if (DEF_VALUES && $native && $native.name !== VALUES) {
    VALUES_BUG = true;
    $default = function values() { return $native.call(this); };
  }
  // Define iterator
  if ((!LIBRARY || FORCED) && (BUGGY || VALUES_BUG || !proto[ITERATOR])) {
    hide(proto, ITERATOR, $default);
  }
  // Plug for library
  Iterators[NAME] = $default;
  Iterators[TAG] = returnThis;
  if (DEFAULT) {
    methods = {
      values: DEF_VALUES ? $default : getMethod(VALUES),
      keys: IS_SET ? $default : getMethod(KEYS),
      entries: $entries
    };
    if (FORCED) for (key in methods) {
      if (!(key in proto)) redefine(proto, key, methods[key]);
    } else $export($export.P + $export.F * (BUGGY || VALUES_BUG), NAME, methods);
  }
  return methods;
};

},{"./_export":29,"./_has":32,"./_hide":33,"./_iter-create":38,"./_iterators":41,"./_library":42,"./_object-gpo":48,"./_redefine":53,"./_set-to-string-tag":54,"./_wks":65}],40:[function(require,module,exports){
module.exports = function (done, value) {
  return { value: value, done: !!done };
};

},{}],41:[function(require,module,exports){
module.exports = {};

},{}],42:[function(require,module,exports){
module.exports = true;

},{}],43:[function(require,module,exports){
'use strict';
// 19.1.2.1 Object.assign(target, source, ...)
var getKeys = require('./_object-keys');
var gOPS = require('./_object-gops');
var pIE = require('./_object-pie');
var toObject = require('./_to-object');
var IObject = require('./_iobject');
var $assign = Object.assign;

// should work with symbols and should have deterministic property order (V8 bug)
module.exports = !$assign || require('./_fails')(function () {
  var A = {};
  var B = {};
  // eslint-disable-next-line no-undef
  var S = Symbol();
  var K = 'abcdefghijklmnopqrst';
  A[S] = 7;
  K.split('').forEach(function (k) { B[k] = k; });
  return $assign({}, A)[S] != 7 || Object.keys($assign({}, B)).join('') != K;
}) ? function assign(target, source) { // eslint-disable-line no-unused-vars
  var T = toObject(target);
  var aLen = arguments.length;
  var index = 1;
  var getSymbols = gOPS.f;
  var isEnum = pIE.f;
  while (aLen > index) {
    var S = IObject(arguments[index++]);
    var keys = getSymbols ? getKeys(S).concat(getSymbols(S)) : getKeys(S);
    var length = keys.length;
    var j = 0;
    var key;
    while (length > j) if (isEnum.call(S, key = keys[j++])) T[key] = S[key];
  } return T;
} : $assign;

},{"./_fails":30,"./_iobject":36,"./_object-gops":47,"./_object-keys":50,"./_object-pie":51,"./_to-object":62}],44:[function(require,module,exports){
// 19.1.2.2 / 15.2.3.5 Object.create(O [, Properties])
var anObject = require('./_an-object');
var dPs = require('./_object-dps');
var enumBugKeys = require('./_enum-bug-keys');
var IE_PROTO = require('./_shared-key')('IE_PROTO');
var Empty = function () { /* empty */ };
var PROTOTYPE = 'prototype';

// Create object with fake `null` prototype: use iframe Object with cleared prototype
var createDict = function () {
  // Thrash, waste and sodomy: IE GC bug
  var iframe = require('./_dom-create')('iframe');
  var i = enumBugKeys.length;
  var lt = '<';
  var gt = '>';
  var iframeDocument;
  iframe.style.display = 'none';
  require('./_html').appendChild(iframe);
  iframe.src = 'javascript:'; // eslint-disable-line no-script-url
  // createDict = iframe.contentWindow.Object;
  // html.removeChild(iframe);
  iframeDocument = iframe.contentWindow.document;
  iframeDocument.open();
  iframeDocument.write(lt + 'script' + gt + 'document.F=Object' + lt + '/script' + gt);
  iframeDocument.close();
  createDict = iframeDocument.F;
  while (i--) delete createDict[PROTOTYPE][enumBugKeys[i]];
  return createDict();
};

module.exports = Object.create || function create(O, Properties) {
  var result;
  if (O !== null) {
    Empty[PROTOTYPE] = anObject(O);
    result = new Empty();
    Empty[PROTOTYPE] = null;
    // add "__proto__" for Object.getPrototypeOf polyfill
    result[IE_PROTO] = O;
  } else result = createDict();
  return Properties === undefined ? result : dPs(result, Properties);
};

},{"./_an-object":19,"./_dom-create":27,"./_enum-bug-keys":28,"./_html":34,"./_object-dps":46,"./_shared-key":55}],45:[function(require,module,exports){
var anObject = require('./_an-object');
var IE8_DOM_DEFINE = require('./_ie8-dom-define');
var toPrimitive = require('./_to-primitive');
var dP = Object.defineProperty;

exports.f = require('./_descriptors') ? Object.defineProperty : function defineProperty(O, P, Attributes) {
  anObject(O);
  P = toPrimitive(P, true);
  anObject(Attributes);
  if (IE8_DOM_DEFINE) try {
    return dP(O, P, Attributes);
  } catch (e) { /* empty */ }
  if ('get' in Attributes || 'set' in Attributes) throw TypeError('Accessors not supported!');
  if ('value' in Attributes) O[P] = Attributes.value;
  return O;
};

},{"./_an-object":19,"./_descriptors":26,"./_ie8-dom-define":35,"./_to-primitive":63}],46:[function(require,module,exports){
var dP = require('./_object-dp');
var anObject = require('./_an-object');
var getKeys = require('./_object-keys');

module.exports = require('./_descriptors') ? Object.defineProperties : function defineProperties(O, Properties) {
  anObject(O);
  var keys = getKeys(Properties);
  var length = keys.length;
  var i = 0;
  var P;
  while (length > i) dP.f(O, P = keys[i++], Properties[P]);
  return O;
};

},{"./_an-object":19,"./_descriptors":26,"./_object-dp":45,"./_object-keys":50}],47:[function(require,module,exports){
exports.f = Object.getOwnPropertySymbols;

},{}],48:[function(require,module,exports){
// 19.1.2.9 / 15.2.3.2 Object.getPrototypeOf(O)
var has = require('./_has');
var toObject = require('./_to-object');
var IE_PROTO = require('./_shared-key')('IE_PROTO');
var ObjectProto = Object.prototype;

module.exports = Object.getPrototypeOf || function (O) {
  O = toObject(O);
  if (has(O, IE_PROTO)) return O[IE_PROTO];
  if (typeof O.constructor == 'function' && O instanceof O.constructor) {
    return O.constructor.prototype;
  } return O instanceof Object ? ObjectProto : null;
};

},{"./_has":32,"./_shared-key":55,"./_to-object":62}],49:[function(require,module,exports){
var has = require('./_has');
var toIObject = require('./_to-iobject');
var arrayIndexOf = require('./_array-includes')(false);
var IE_PROTO = require('./_shared-key')('IE_PROTO');

module.exports = function (object, names) {
  var O = toIObject(object);
  var i = 0;
  var result = [];
  var key;
  for (key in O) if (key != IE_PROTO) has(O, key) && result.push(key);
  // Don't enum bug & hidden keys
  while (names.length > i) if (has(O, key = names[i++])) {
    ~arrayIndexOf(result, key) || result.push(key);
  }
  return result;
};

},{"./_array-includes":20,"./_has":32,"./_shared-key":55,"./_to-iobject":60}],50:[function(require,module,exports){
// 19.1.2.14 / 15.2.3.14 Object.keys(O)
var $keys = require('./_object-keys-internal');
var enumBugKeys = require('./_enum-bug-keys');

module.exports = Object.keys || function keys(O) {
  return $keys(O, enumBugKeys);
};

},{"./_enum-bug-keys":28,"./_object-keys-internal":49}],51:[function(require,module,exports){
exports.f = {}.propertyIsEnumerable;

},{}],52:[function(require,module,exports){
module.exports = function (bitmap, value) {
  return {
    enumerable: !(bitmap & 1),
    configurable: !(bitmap & 2),
    writable: !(bitmap & 4),
    value: value
  };
};

},{}],53:[function(require,module,exports){
module.exports = require('./_hide');

},{"./_hide":33}],54:[function(require,module,exports){
var def = require('./_object-dp').f;
var has = require('./_has');
var TAG = require('./_wks')('toStringTag');

module.exports = function (it, tag, stat) {
  if (it && !has(it = stat ? it : it.prototype, TAG)) def(it, TAG, { configurable: true, value: tag });
};

},{"./_has":32,"./_object-dp":45,"./_wks":65}],55:[function(require,module,exports){
var shared = require('./_shared')('keys');
var uid = require('./_uid');
module.exports = function (key) {
  return shared[key] || (shared[key] = uid(key));
};

},{"./_shared":56,"./_uid":64}],56:[function(require,module,exports){
var global = require('./_global');
var SHARED = '__core-js_shared__';
var store = global[SHARED] || (global[SHARED] = {});
module.exports = function (key) {
  return store[key] || (store[key] = {});
};

},{"./_global":31}],57:[function(require,module,exports){
var toInteger = require('./_to-integer');
var defined = require('./_defined');
// true  -> String#at
// false -> String#codePointAt
module.exports = function (TO_STRING) {
  return function (that, pos) {
    var s = String(defined(that));
    var i = toInteger(pos);
    var l = s.length;
    var a, b;
    if (i < 0 || i >= l) return TO_STRING ? '' : undefined;
    a = s.charCodeAt(i);
    return a < 0xd800 || a > 0xdbff || i + 1 === l || (b = s.charCodeAt(i + 1)) < 0xdc00 || b > 0xdfff
      ? TO_STRING ? s.charAt(i) : a
      : TO_STRING ? s.slice(i, i + 2) : (a - 0xd800 << 10) + (b - 0xdc00) + 0x10000;
  };
};

},{"./_defined":25,"./_to-integer":59}],58:[function(require,module,exports){
var toInteger = require('./_to-integer');
var max = Math.max;
var min = Math.min;
module.exports = function (index, length) {
  index = toInteger(index);
  return index < 0 ? max(index + length, 0) : min(index, length);
};

},{"./_to-integer":59}],59:[function(require,module,exports){
// 7.1.4 ToInteger
var ceil = Math.ceil;
var floor = Math.floor;
module.exports = function (it) {
  return isNaN(it = +it) ? 0 : (it > 0 ? floor : ceil)(it);
};

},{}],60:[function(require,module,exports){
// to indexed object, toObject with fallback for non-array-like ES3 strings
var IObject = require('./_iobject');
var defined = require('./_defined');
module.exports = function (it) {
  return IObject(defined(it));
};

},{"./_defined":25,"./_iobject":36}],61:[function(require,module,exports){
// 7.1.15 ToLength
var toInteger = require('./_to-integer');
var min = Math.min;
module.exports = function (it) {
  return it > 0 ? min(toInteger(it), 0x1fffffffffffff) : 0; // pow(2, 53) - 1 == 9007199254740991
};

},{"./_to-integer":59}],62:[function(require,module,exports){
// 7.1.13 ToObject(argument)
var defined = require('./_defined');
module.exports = function (it) {
  return Object(defined(it));
};

},{"./_defined":25}],63:[function(require,module,exports){
// 7.1.1 ToPrimitive(input [, PreferredType])
var isObject = require('./_is-object');
// instead of the ES6 spec version, we didn't implement @@toPrimitive case
// and the second argument - flag - preferred type is a string
module.exports = function (it, S) {
  if (!isObject(it)) return it;
  var fn, val;
  if (S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it))) return val;
  if (typeof (fn = it.valueOf) == 'function' && !isObject(val = fn.call(it))) return val;
  if (!S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it))) return val;
  throw TypeError("Can't convert object to primitive value");
};

},{"./_is-object":37}],64:[function(require,module,exports){
var id = 0;
var px = Math.random();
module.exports = function (key) {
  return 'Symbol('.concat(key === undefined ? '' : key, ')_', (++id + px).toString(36));
};

},{}],65:[function(require,module,exports){
var store = require('./_shared')('wks');
var uid = require('./_uid');
var Symbol = require('./_global').Symbol;
var USE_SYMBOL = typeof Symbol == 'function';

var $exports = module.exports = function (name) {
  return store[name] || (store[name] =
    USE_SYMBOL && Symbol[name] || (USE_SYMBOL ? Symbol : uid)('Symbol.' + name));
};

$exports.store = store;

},{"./_global":31,"./_shared":56,"./_uid":64}],66:[function(require,module,exports){
var classof = require('./_classof');
var ITERATOR = require('./_wks')('iterator');
var Iterators = require('./_iterators');
module.exports = require('./_core').getIteratorMethod = function (it) {
  if (it != undefined) return it[ITERATOR]
    || it['@@iterator']
    || Iterators[classof(it)];
};

},{"./_classof":21,"./_core":23,"./_iterators":41,"./_wks":65}],67:[function(require,module,exports){
var anObject = require('./_an-object');
var get = require('./core.get-iterator-method');
module.exports = require('./_core').getIterator = function (it) {
  var iterFn = get(it);
  if (typeof iterFn != 'function') throw TypeError(it + ' is not iterable!');
  return anObject(iterFn.call(it));
};

},{"./_an-object":19,"./_core":23,"./core.get-iterator-method":66}],68:[function(require,module,exports){
'use strict';
var addToUnscopables = require('./_add-to-unscopables');
var step = require('./_iter-step');
var Iterators = require('./_iterators');
var toIObject = require('./_to-iobject');

// 22.1.3.4 Array.prototype.entries()
// 22.1.3.13 Array.prototype.keys()
// 22.1.3.29 Array.prototype.values()
// 22.1.3.30 Array.prototype[@@iterator]()
module.exports = require('./_iter-define')(Array, 'Array', function (iterated, kind) {
  this._t = toIObject(iterated); // target
  this._i = 0;                   // next index
  this._k = kind;                // kind
// 22.1.5.2.1 %ArrayIteratorPrototype%.next()
}, function () {
  var O = this._t;
  var kind = this._k;
  var index = this._i++;
  if (!O || index >= O.length) {
    this._t = undefined;
    return step(1);
  }
  if (kind == 'keys') return step(0, index);
  if (kind == 'values') return step(0, O[index]);
  return step(0, [index, O[index]]);
}, 'values');

// argumentsList[@@iterator] is %ArrayProto_values% (9.4.4.6, 9.4.4.7)
Iterators.Arguments = Iterators.Array;

addToUnscopables('keys');
addToUnscopables('values');
addToUnscopables('entries');

},{"./_add-to-unscopables":18,"./_iter-define":39,"./_iter-step":40,"./_iterators":41,"./_to-iobject":60}],69:[function(require,module,exports){
// 19.1.3.1 Object.assign(target, source)
var $export = require('./_export');

$export($export.S + $export.F, 'Object', { assign: require('./_object-assign') });

},{"./_export":29,"./_object-assign":43}],70:[function(require,module,exports){
var $export = require('./_export');
// 19.1.2.4 / 15.2.3.6 Object.defineProperty(O, P, Attributes)
$export($export.S + $export.F * !require('./_descriptors'), 'Object', { defineProperty: require('./_object-dp').f });

},{"./_descriptors":26,"./_export":29,"./_object-dp":45}],71:[function(require,module,exports){
'use strict';
var $at = require('./_string-at')(true);

// 21.1.3.27 String.prototype[@@iterator]()
require('./_iter-define')(String, 'String', function (iterated) {
  this._t = String(iterated); // target
  this._i = 0;                // next index
// 21.1.5.2.1 %StringIteratorPrototype%.next()
}, function () {
  var O = this._t;
  var index = this._i;
  var point;
  if (index >= O.length) return { value: undefined, done: true };
  point = $at(O, index);
  this._i += point.length;
  return { value: point, done: false };
});

},{"./_iter-define":39,"./_string-at":57}],72:[function(require,module,exports){
require('./es6.array.iterator');
var global = require('./_global');
var hide = require('./_hide');
var Iterators = require('./_iterators');
var TO_STRING_TAG = require('./_wks')('toStringTag');

var DOMIterables = ('CSSRuleList,CSSStyleDeclaration,CSSValueList,ClientRectList,DOMRectList,DOMStringList,' +
  'DOMTokenList,DataTransferItemList,FileList,HTMLAllCollection,HTMLCollection,HTMLFormElement,HTMLSelectElement,' +
  'MediaList,MimeTypeArray,NamedNodeMap,NodeList,PaintRequestList,Plugin,PluginArray,SVGLengthList,SVGNumberList,' +
  'SVGPathSegList,SVGPointList,SVGStringList,SVGTransformList,SourceBufferList,StyleSheetList,TextTrackCueList,' +
  'TextTrackList,TouchList').split(',');

for (var i = 0; i < DOMIterables.length; i++) {
  var NAME = DOMIterables[i];
  var Collection = global[NAME];
  var proto = Collection && Collection.prototype;
  if (proto && !proto[TO_STRING_TAG]) hide(proto, TO_STRING_TAG, NAME);
  Iterators[NAME] = Iterators.Array;
}

},{"./_global":31,"./_hide":33,"./_iterators":41,"./_wks":65,"./es6.array.iterator":68}]},{},[1]);
